runtime/commonMain/src/kotlinx/serialization/json/internal/JsonReader.kt

```diff
@@ -165,94 +164,86 @@ internal class JsonReader(private val source: String) {
-            when (tc) {
-                TC_WS -> curPos++ // skip whitespace
+        var currentPosition = currentPosition
+        while (currentPosition < source.length) {
+            val ch = source[currentPosition]
+            when (val tc = charToTokenClass(ch)) {
+                TC_WS -> currentPosition++ // skip whitespace
                 TC_OTHER -> {
-                    nextLiteral(source, curPos)
+                    nextLiteral(source, currentPosition)
                     return
                 }
                 TC_STRING -> {
-                    nextString(source, curPos)
+                    nextString(source, currentPosition)
                     return
                 }
                 else -> {
-                    this.tokenPosition = curPos
+                    this.tokenPosition = currentPosition
                     this.tokenClass = tc
-                    this.currentPosition = curPos + 1
+                    this.currentPosition = currentPosition + 1
                     return
                 }
             }
         }
+
+        tokenPosition = currentPosition
+        tokenClass = TC_EOF
     }
 
     private fun nextLiteral(source: String, startPos: Int) {
         tokenPosition = startPos
         offset = startPos
-        var curPos = startPos
-        val maxLen = source.length
-        while (true) {
-            curPos++
-            if (curPos >= maxLen || charToTokenClass(source[curPos]) != TC_OTHER) break
+        var currentPosition = startPos
+        while (currentPosition < source.length && charToTokenClass(source[currentPosition]) == TC_OTHER) {
+            currentPosition++
         }
-        this.currentPosition = curPos
-        length = curPos - offset
+        this.currentPosition = currentPosition
+        length = currentPosition - offset
         tokenClass = if (rangeEquals(source, offset, length, NULL)) TC_NULL else TC_OTHER
     }
 
-    private fun nextString(source: String, startPos: Int) {
-        tokenPosition = startPos
+    private fun nextString(source: String, startPosition: Int) {
+        tokenPosition = startPosition
         length = 0 // in buffer
-        var curPos = startPos + 1
-        var lastPos = curPos
-        val maxLen = source.length
-        parse@ while (true) {
-            if (curPos >= maxLen) fail(curPos, "Unexpected end in string")
-            if (source[curPos] == STRING) {
-                break@parse
-            } else if (source[curPos] == STRING_ESC) {
-                appendRange(source, lastPos, curPos)
-                val newPos = appendEsc(source, curPos + 1)
-                curPos = newPos
-                lastPos = newPos
+        var currentPosition = startPosition + 1
+        var lastPosition = currentPosition
+        val length = source.length
+        while (source[currentPosition] != STRING) {
+            if (currentPosition >= length) fail("Unexpected end in string", currentPosition)
```

Comment:
"Unexpected end **in** string" or "**of** string"? I know this is not your text, but since we are fixing exceptions messages anyway...